noir-projects/noir-contracts/contracts/auth_contract/src/test/main.nr

```diff
@@ -0,0 +1,119 @@
+    admin_sets_authorized();
+
+    let authorized_is_not_yet_effective = || {
+        env.impersonate(to_authorize);
+        let authorized: AztecAddress = env.call_public(Auth::at(auth_contract_address).get_authorized());
+        assert_eq(authorized, AztecAddress::zero());
+
+        env.assert_private_call_fails(Auth::at(auth_contract_address).do_private_authorized_thing());
+    };
+    authorized_is_not_yet_effective();
+
+    let authorized_becomes_effective_after_delay = || {
+        env.impersonate(to_authorize);
+
+        // We advance block by 4, because we initially advanced block by one after setting the value. See below comment for explanation.
+        env.advance_block_by(4);
+        let authorized: AztecAddress = env.call_public(Auth::at(auth_contract_address).get_authorized());
+        assert_eq(authorized, to_authorize);
+
+        let authorized_in_private: AztecAddress = env.call_private(Auth::at(auth_contract_address).get_authorized_in_private());
+        assert_eq(authorized_in_private, AztecAddress::zero());
+
+        // We need to always advance the block one more time to get the current value in private, compared to the value in public.
+        // To see why let's see this diagram.
+        // When we make a change in public, lets say we are at block 2 (building a tx to be included in block 2), which means the latest committed block is block 1.
+        // Thus, the value change will be set to block 7 (2 + 5).
+        // If we now advance our env by 5 blocks, we will be at block 7 (building a tx to be included in block 7), which means the latest committed block is block 6.
+        // Reading the value in public will work, because it will use the current block (7), and the current block is the block of change; but
+        // if we try to create a historical proof, we do not have access to block 7 yet, and have to build the proof off of block 6, but at this time, the value change will not have
+        // taken place yet, therefore we need to be at block 8 (building a tx to be included in block 8), for the historical proof to work, as it will have access to the full block 7
+        // where the value change takes effect.
+        // Note: We do not see this behavior in the e2e tests because setting the value inplicitly advances the block number by 1.
+        //                              1     2     3     4     5     6     7     8     9
+        //                              |     |     |     |     |     |     |     |     |
+        //                                 ^
+        //                    value change scheduled here
+        //                                                                ^
+        //                                   get_authorized() (public) called here with block_number = 7
+        //                                                                      ^
+        //                                        get_authorized() (private) called here with block_number = 8
```

Comment:
This is a very good explanation, but it ultimately shows that the current API is flawed. Users should not have to be exposed to this level of detail for something as simple as this. We should at least have `call_private_at`.